### PR TITLE
[i2c] Consolidate transaction boundary tracking in a bus monitor

### PR DESCRIPTION
**Pull request by a-will**
Prepare for multi-controller functionality by creating a bus monitor that tracks transaction boundaries, detects Start and Stop symbols, and provides clock timeout functions. Move "bus free" detection to the bus monitor, and block the controller FSM from beginning transmissions unless the bus monitor gives the OK.

Add an optional SMBus bus timeout, which checks if SCL is held low for too long, and make it mux with the target stretch timeout, since it doesn't make much sense for them to both be active. In the general bus timeout case, it doesn't matter which entity is pulling SCL low. In the target stretch timeout case, the controller's contribution to SCL low time is ignored.

Add a control bit for selecting between single-controller and multi-controller monitoring mode. For the single-controller case, the bus monitor starts with the bus assumed to be free, and it only blocks the controller FSM from transmitting during the t_buf window after a Stop condition. For the multi-controller case, the bus monitor begins in a bus busy state, and the controller FSM must wait for at least the idle timeout (which _must_ be set in HOST_TIMEOUT_CTRL) before it is allowed to transmit.

Remove the "SDA is low in Idle" check from the SDA interference interrupt. There are various single-controller scenarios where this may trigger uselessly, and it isn't compatible with multi-controller scenarios.

Finally, fix a couple miscellaneous issues for DV:
- Eliminate ACQ_FIFO_NEXT_DATA from auto CSR checks (needs i2c traffic)
- Prevent address matching when the mask is 0

### CODE REVIEW

File: hw/ip/i2c/rtl/i2c_controller_fsm.sv [hunk 1]
```diff
@@ -828,15 +828,13 @@ module i2c_controller_fsm import i2c_pkg::*;
...
-          load_tcount = 1'b1;
-          tcount_sel = tHoldStop;
           log_stop = 1'b1;
         end
       end
       // HoldStop: SDA and SCL are released
       HoldStop : begin
         en_sda_interf_det = 1'b1;
-        if (tcount_q == 20'd1) begin
+        if (scl_i && sda_i) begin
```
Comment (a-will) [inline hw/ip/i2c/rtl/i2c_controller_fsm.sv:837]: The bus monitor now counts out the bus free time, so we simply wait for the signals to finish rising here (and keep the SDA interference check enabled meanwhile). If the bus gets stuck with SDA low, software can reset the FSM by setting CTRL.ENABLEHOST to 0.

File: hw/ip/i2c/data/i2c.hjson [hunk 2]
```diff
@@ -664,19 +679,39 @@
...
       desc: '''
-            I2C clock stretching timeout control.
-            This is used in I2C host mode to detect whether a connected target is stretching for too long.
+            I2C clock stretching and bus timeout control.
+            The behavior of this feature depends on the value of TIMEOUT_CTRL.BUS_TIMEOUT_EN.
+            If the bus timeout is disabled, this is used in I2C controller mode to detect whether a connected target is stretching a single low time beyond the timeout value.
+            When the bus timeout feature is disabled, this feature is more informative and doesn't do more than assert the "stretch_timeout" interrupt.
+
+            If the bus timeout feature is enabled, it is used to detect whether the clock has been held low for too long instead, inclusive of the controller's clock low time.
+            In an SMBus context, the VAL programmed should be tTIMEOUT:MIN, and EN and BUS_TIMEOUT_EN should be 1.
```
Comment (a-will) [inline hw/ip/i2c/data/i2c.hjson:688]: This feature isn't totally complete yet, since the target FSM's responses aren't in this PR. Those will come in a follow-up.

File: sw/device/tests/i2c_target_smbus_arp_test.c [hunk 3]
```diff
@@ -297,11 +297,11 @@ static status_t test_init(void) {
...
   TRY(i2c_testutils_set_speed(&i2c, kDifI2cSpeedStandard));
 
-  // 20 ms stretch timeout. The upper limit of support is ~1 GHz for the IP, so
-  // OK to make the frequency a uint32_t.
-  uint32_t kStretchCycles = ((uint32_t)kClockFreqPeripheralHz / (1000 / 20));
-  TRY(dif_i2c_enable_clock_stretching_timeout(&i2c, kDifToggleEnabled,
-                                              kStretchCycles));
+  // 25 ms bus timeout. The upper limit of support is ~1 GHz for the IP, so OK
+  // to make the frequency a uint32_t.
+  uint32_t kBusTimeoutCycles = ((uint32_t)kClockFreqPeripheralHz / (1000 / 20));
```
Comment (a-will) [inline sw/device/tests/i2c_target_smbus_arp_test.c:302]: Whoops, forgot to actually change it to 25 ms, hehe.

File: sw/device/tests/i2c_target_smbus_arp_test.c [hunk 4]
```diff
@@ -319,6 +319,7 @@ static status_t test_init(void) {
   TRY(dif_i2c_set_target_watermarks(&i2c, /*tx_level=*/0,
                                     I2C_PARAM_ACQ_FIFO_DEPTH));
   TRY(dif_i2c_set_device_id(&i2c, &kSmbusArpId, NULL));
+  TRY(dif_i2c_multi_controller_monitor_set_enabled(&i2c, kDifToggleEnabled));
```
Comment (a-will) [inline sw/device/tests/i2c_target_smbus_arp_test.c:322]: In this test, the controller module isn't actually enabled, but we put the monitor in multi-controller mode to get a tiny bit of coverage on the bus monitor.

**Review by hcallahan-lowrisc — APPROVED** [review]
LGTM. A bit tricky to review the new features without them really being exercised, but on inspection all the logic seems sensible to me. I certainly agree that it feels natural to bring the bus_monitor logic into its own module, so I am pleased to see that change.
Just a few small doc nits that can be swept up into one of the later PRs if desired.

File: hw/ip/i2c/data/i2c.hjson [hunk 5]
```diff
@@ -272,6 +272,21 @@
...
+          desc: '''
+                Enable the bus monitor in multi-controller mode.
+
+                If a 0->1 transition happens while !!CTRL.ENABLEHOST and !!CTRL.ENABLETARGET are both 0, the bus monitor will enable and begin in the "bus busy" state.
+                To transition to a bus free state, !!HOST_TIMEOUT_CTRL must be nonzero, so the bus monitor may count out idle cycles to confirm the freedom to transmit.
+                In addition, the bus monitor will track whether the bus is free based on the enabled timeouts and detected Stop symbols.
+                For multi-controller mode, ensure !!CTRL.MULTI_CONTROLLER_MONITOR_EN becomes 1 no later than !!CTRL.ENABLEHOST or !!CTRL.ENABLETARGET.
+
+                When 0, the bus monitor will report that the bus is always free, so the controller FSM is never blocked from transmitting.
+                If !!CTRL.MULTI_CONTROLLER_MONITOR_EN is set after !!CTRL.ENABLEHOST or !!CTRL.ENABLETARGET, the bus monitor will begin in the "bus free" state instead.
```
[the remark this answers is not in the document]
Comment (a-will) [reply hw/ip/i2c/data/i2c.hjson:287]: Ah, the last sentence is kind of a warning, with an indication of how it would violate the spec. In multi-controller mode, the bus monitor must assume the bus is busy at initial connection, and achieving that requires this bit being set to 1 first (or coincident with the others).

In the future doc PR, I'll try to reword it to make it clearer.

File: hw/ip/i2c/rtl/i2c_bus_monitor.sv [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+  input                            rst_ni,
+
+  input                            scl_i,
+  input                            sda_i,
+
+  input                            controller_enable_i,
+  input                            multi_controller_enable_i,
+  input                            target_enable_i,
+  input                            target_idle_i,
+  input [12:0]                     thd_dat_i,              // Data hold time(?XX ns)
```
[the remark this answers is not in the document]
Comment (a-will) [reply hw/ip/i2c/rtl/i2c_bus_monitor.sv:20]: That means "dozens or possibly a hundred or so nanoseconds", haha

File: hw/ip/i2c/data/i2c.hjson [hunk 7]
```diff
@@ -664,19 +679,39 @@
...
       fields: [
-        { bits: "30:0"
+        { bits: "29:0"
           name: "VAL"
           desc: "Clock stretching timeout value (in units of input clock frequency)"
         }
+        { bits: "30"
+          name: "BUS_TIMEOUT_EN"
+          desc: '''
+                Enable bus timeout feature.
```
[the remark this answers is not in the document]
Comment (a-will) [reply hw/ip/i2c/data/i2c.hjson:700]: Perhaps I should change it to be an enum that selects the particular mode instead.

File: hw/ip/i2c/rtl/i2c_bus_monitor.sv (hunk 6 above)
[the remark this answers is not in the document]
Comment (a-will) [reply hw/ip/i2c/rtl/i2c_bus_monitor.sv]: Changed to be a bit more explanatory.

File: hw/ip/i2c/data/i2c.hjson (hunk 7 above)
[the remark this answers is not in the document]
Comment (a-will) [reply hw/ip/i2c/data/i2c.hjson]: Changed to enum. Hopefully it makes more sense now!

File: hw/ip/i2c/data/i2c.hjson
[the remark this answers is not in the document]
Comment (a-will) [reply hw/ip/i2c/data/i2c.hjson:1]: Done.

File: hw/ip/i2c/data/i2c.hjson (hunk 5 above)
[the remark this answers is not in the document]
Comment (a-will) [reply hw/ip/i2c/data/i2c.hjson]: Or this one... I edited it, and hopefully it's a little bit clearer.

**Comment by a-will** [comment]
I added a little tweak to reggen: For the C header output, it was skipping enums that were only one-bit wide, and I stopped that. I wanted those values in my header! :)

**Comment by hcallahan-lowrisc** [comment]
LGTM! Thanks for tweaking the docs, and fixing the reggen bug!

**Comment by a-will** [comment]
I'm going to merge this so it can get in for regressions. If there are fixups needed, we'll get them in the next PR.
